lib/beaker/hypervisor.rb

```diff
@@ -57,6 +58,26 @@ def cleanup
       nil
     end
 
+    DEFAULT_CONNECTION_PREFERENCE = [:ip, :vmhostname, :hostname]
+    # SSH connection method preference. Can be overwritten by hypervisor to change the order
+    def connection_preference(host)
+      DEFAULT_CONNECTION_PREFERENCE
+    end
+
+    def self.set_ssh_connection_preference(hosts_to_provision, hypervisor)
+      hosts_to_provision.each do |host|
+        ssh_methods = hypervisor.connection_preference(host)
+        if host[:ssh_preference]
+          # If user has provided ssh_connection_preference in hosts file then concat the preference provided by hypervisor
+          # Followed by concatenating the default preference and keeping the unique once
+          ssh_methods = (host[:ssh_preference] + (ssh_methods + DEFAULT_CONNECTION_PREFERENCE)).uniq
+        else
+          ssh_methods = (ssh_methods + DEFAULT_CONNECTION_PREFERENCE).uniq
+        end
+        host[:ssh_connection_preference] = ssh_methods
```

Comment:
...and then change this line to `host[:ssh_connection_preference] = ssh_methods.uniq`